project_functional_block/models/project_functional_block.py

```diff
@@ -0,0 +1,152 @@
+#    published by the Free Software Foundation, either version 3 of the
+#    License, or (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from openerp import models, fields, api
+
+
+class ProjectFunctionalBlock(models.Model):
+    _name = 'project.functional.block'
+    _description = 'Functional block to organize projects tasks'
+    _order = 'complete_name'
+    _rec_name = 'complete_name'
+
+    # Columns
+    name = fields.Char('Name', size=64, required=True, translate=True)
+    code = fields.Char('Code', size=8)
+    # TODO: this field doesn't work with multi-languages
+    complete_name = fields.Char(string="Complete Name", index=True)
+    description = fields.Text('Description', translate=True)
+    parent_id = fields.Many2one(
+        'project.functional.block', 'Parent Block',
+        ondelete='cascade', index=True)
+    tasks_count = fields.Integer(string='Tasks Count', compute='_count_tasks')
+    child_fblocks_count = fields.Integer(
+        string='Child Blocks Count', compute='_count_child_fblocks')
+
+    @api.multi
+    def _update_complete_name(self):
+        """
+        Complete Name includes parent's name and name of this functional block
+        itself.
+        """
+        todo_fblocks = self.search([('id', 'child_of', self.ids)],
+                                   order='complete_name')
+        for func_block in todo_fblocks:
+            complete_name = func_block.name
+            parent = func_block.parent_id
+            while parent:
+                complete_name = parent.name + ' / ' + complete_name
+                parent = parent.parent_id
+            func_block.complete_name = complete_name
+
+    @api.model
+    def create(self, vals):
+        new_fblock = super(ProjectFunctionalBlock, self).create(vals)
+        # Update complete_name for this record
+        if 'updating_complete_name' not in self._context:
+            ctx = self._context.copy()
+            ctx.update({'updating_complete_name': True})
+            new_fblock.with_context(ctx)._update_complete_name()
+        return new_fblock
+
+    @api.multi
+    def write(self, vals):
+        res = super(ProjectFunctionalBlock, self).write(vals)
+        # Update complete_name for these records and child records
+        if 'updating_complete_name' not in self._context:
+            ctx = self._context.copy()
+            ctx.update({'updating_complete_name': True})
+            self.with_context(ctx)._update_complete_name()
+        return res
```

Comment:
I feel that the logic in create() and writ()e could be replaced by a converting complete_name into a computed field.
But I won't pick on that, just add it to the Known Issues/Roadmap section of the README please.